File: app/src/main/java/com/example/editor/EditorManager.java
package com.example.editor;

import com.example.filesystem.LocalFileSystem;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EditorManager {

    private static volatile EditorManager INSTANCE;

    private final List<EditorTab> tabs = new ArrayList<>();
    private final Map<String, UndoManager> undoManagers = new HashMap<>();
    private int activeTabIndex = -1;
    private final LocalFileSystem fileSystem;

    private EditorManager() {
        this.fileSystem = LocalFileSystem.getInstance();
    }

    public static EditorManager getInstance() {
        if (INSTANCE == null) {
            synchronized (EditorManager.class) {
                if (INSTANCE == null) {
                    INSTANCE = new EditorManager();
                }
            }
        }
        return INSTANCE;
    }

    public List<EditorTab> getTabs() {
        return tabs;
    }

    public EditorTab getActiveTab() {
        if (activeTabIndex >= 0 && activeTabIndex < tabs.size()) {
            return tabs.get(activeTabIndex);
        }
        return null;
    }

    public int getActiveTabIndex() {
        return activeTabIndex;
    }

    public void setActiveTabIndex(int index) {
        if (index >= 0 && index < tabs.size()) {
            this.activeTabIndex = index;
        } else if (tabs.isEmpty()) {
            this.activeTabIndex = -1;
        }
    }

    public EditorTab openFile(File file) throws Exception {
        if (file == null || !file.exists() || !file.isFile()) {
            throw new IllegalArgumentException("File does not exist or is not a valid file");
        }

        String path = file.getAbsolutePath();
        // Check if already open
        for (int i = 0; i < tabs.size(); i++) {
            if (tabs.get(i).getFilePath().equals(path)) {
                activeTabIndex = i;
                return tabs.get(i);
            }
        }

        String content = fileSystem.readFileToString(file);
        EditorTab tab = new EditorTab(path, file.getName(), content, null);
        tabs.add(tab);
        activeTabIndex = tabs.size() - 1;

        UndoManager undoMgr = new UndoManager();
        undoMgr.pushState(content);
        undoManagers.put(path, undoMgr);

        return tab;
    }

    public void updateActiveTabContent(String newContent) {
        EditorTab tab = getActiveTab();
        if (tab != null) {
            tab.updateContent(newContent);
            UndoManager undoMgr = undoManagers.get(tab.getFilePath());
            if (undoMgr != null) {
                undoMgr.pushState(newContent);
            }
        }
    }

    public void saveActiveTab() throws Exception {
        EditorTab tab = getActiveTab();
        if (tab != null && tab.isModified()) {
            fileSystem.writeStringToFile(tab.getFile(), tab.getContent());
            tab.markSaved();
        }
    }

    public void saveAllTabs() throws Exception {
        for (EditorTab tab : tabs) {
            if (tab.isModified()) {
                fileSystem.writeStringToFile(tab.getFile(), tab.getContent());
                tab.markSaved();
            }
        }
    }

    public void undoActiveTab() {
        EditorTab tab = getActiveTab();
        if (tab != null) {
            UndoManager undoMgr = undoManagers.get(tab.getFilePath());
            if (undoMgr != null && undoMgr.canUndo()) {
                String previous = undoMgr.undo(tab.getContent());
                tab.updateContent(previous);
            }
        }
    }

    public void redoActiveTab() {
        EditorTab tab = getActiveTab();
        if (tab != null) {
            UndoManager undoMgr = undoManagers.get(tab.getFilePath());
            if (undoMgr != null && undoMgr.canRedo()) {
                String next = undoMgr.redo(tab.getContent());
                tab.updateContent(next);
            }
        }
    }

    public void closeTab(int index) {
        if (index >= 0 && index < tabs.size()) {
            EditorTab removed = tabs.remove(index);
            undoManagers.remove(removed.getFilePath());
            if (activeTabIndex >= tabs.size()) {
                activeTabIndex = tabs.size() - 1;
            }
        }
    }

    public void closeAllTabs() {
        tabs.clear();
        undoManagers.clear();
        activeTabIndex = -1;
    }

    public void closeOtherTabs(int keepIndex) {
        if (keepIndex >= 0 && keepIndex < tabs.size()) {
            EditorTab keepTab = tabs.get(keepIndex);
            tabs.clear();
            tabs.add(keepTab);
            activeTabIndex = 0;
        }
    }
}

File: app/src/main/java/com/droidcode/editor/EditorManager.kt
package com.droidcode.editor

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.setValue
import com.droidcode.filesystem.LocalFileSystem
import java.io.File

class EditorManager private constructor() {

    val tabs = mutableStateListOf<EditorTab>()
    private val undoManagers = mutableMapOf<String, UndoManager>()

    var activeTabIndex: Int by mutableIntStateOf(-1)

    private val fileSystem = LocalFileSystem.getInstance()

    val activeTab: EditorTab?
        get() = if (activeTabIndex in 0 until tabs.size) tabs[activeTabIndex] else null

    fun openFile(file: File): EditorTab {
        require(file.exists() && file.isFile) { "File does not exist or is not a valid file: ${file.path}" }

        val path = file.absolutePath
        // Check if already open
        tabs.forEachIndexed { index, tab ->
            if (tab.filePath == path) {
                activeTabIndex = index
                return tab
            }
        }

        val viewerType = EditorTab.detectViewerType(file)
        val content = if (viewerType == FileViewerType.TEXT) {
            try {
                fileSystem.readFileToString(file) ?: ""
            } catch (e: Exception) {
                ""
            }
        } else {
            ""
        }

        val tab = EditorTab(
            filePath = path,
            fileName = file.name,
            initialContent = content,
            initialViewerType = viewerType
        )
        tabs.add(tab)
        activeTabIndex = tabs.size - 1

        val undoMgr = UndoManager()
        undoMgr.pushState(content)
        undoManagers[path] = undoMgr

        return tab
    }

    fun updateActiveTabContent(newContent: String) {
        val tab = activeTab ?: return
        tab.updateContent(newContent)
        undoManagers[tab.filePath]?.pushState(newContent)
    }

    fun saveActiveTab() {
        val tab = activeTab ?: return
        saveTab(tab)
    }

    fun saveTab(tab: EditorTab) {
        if (tab.isModified) {
            fileSystem.writeStringToFile(tab.file, tab.content)
            tab.markSaved()
        }
    }

    fun saveAllTabs() {
        tabs.forEach { tab ->
            if (tab.isModified) {
                fileSystem.writeStringToFile(tab.file, tab.content)
                tab.markSaved()
            }
        }
    }

    fun canUndoActiveTab(): Boolean {
        val tab = activeTab ?: return false
        val undoMgr = undoManagers[tab.filePath] ?: return false
        return undoMgr.canUndo()
    }

    fun canRedoActiveTab(): Boolean {
        val tab = activeTab ?: return false
        val undoMgr = undoManagers[tab.filePath] ?: return false
        return undoMgr.canRedo()
    }

    fun undoActiveTab() {
        val tab = activeTab ?: return
        val undoMgr = undoManagers[tab.filePath] ?: return
        if (undoMgr.canUndo()) {
            val previous = undoMgr.undo(tab.content)
            tab.updateContent(previous)
        }
    }

    fun redoActiveTab() {
        val tab = activeTab ?: return
        val undoMgr = undoManagers[tab.filePath] ?: return
        if (undoMgr.canRedo()) {
            val next = undoMgr.redo(tab.content)
            tab.updateContent(next)
        }
    }

    fun closeTab(index: Int) {
        if (index in 0 until tabs.size) {
            val removed = tabs.removeAt(index)
            undoManagers.remove(removed.filePath)
            if (activeTabIndex >= tabs.size) {
                activeTabIndex = tabs.size - 1
            }
        }
    }

    fun closeAllTabs() {
        tabs.clear()
        undoManagers.clear()
        activeTabIndex = -1
    }

    fun reloadActiveTab() {
        val tab = activeTab ?: return
        if (tab.file.exists()) {
            val fresh = fileSystem.readFileToString(tab.file) ?: ""
            tab.forceOpenAsText(fresh)
            tab.markSaved()
        }
    }

    fun closeTabsToRight(fromIndex: Int) {
        if (fromIndex in 0 until tabs.size) {
            val lastIdx = tabs.size - 1
            for (i in lastIdx downTo (fromIndex + 1)) {
                closeTab(i)
            }
        }
    }

    fun closeOtherTabs(keepIndex: Int) {
        if (keepIndex in 0 until tabs.size) {
            val keepTab = tabs[keepIndex]
            tabs.clear()
            tabs.add(keepTab)
            activeTabIndex = 0
        }
    }

    companion object {
        @Volatile
        private var instance: EditorManager? = null

        @JvmStatic
        fun getInstance(): EditorManager {
            return instance ?: synchronized(this) {
                instance ?: EditorManager().also { instance = it }
            }
        }
    }
}
